core/src/main/java/org/infinispan/marshall/protostream/impl/MarshallableDeque.java

```diff
@@ -0,0 +1,94 @@
+package org.infinispan.marshall.protostream.impl;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Queue;
+
+import org.infinispan.commons.marshall.ProtoStreamTypeIds;
+import org.infinispan.protostream.annotations.ProtoFactory;
+import org.infinispan.protostream.annotations.ProtoField;
+import org.infinispan.protostream.annotations.ProtoTypeId;
+
+/**
+ * A wrapper for a {@link Deque} of objects whose type is unknown until runtime. This is equivalent to utilising a
+ * <code>Deque<MarshallableObject></code> without the overhead of creating a {@link MarshallableObject} per
+ * entry.
+ *
+ * @author Ryan Emerson
+ * @since 16.0
+ */
+@ProtoTypeId(ProtoStreamTypeIds.MARSHALLABLE_DEQUE)
+public class MarshallableDeque<T> extends AbstractMarshallableCollection<T> {
+
+   /**
+    * @param queue the {@link Queue} to be wrapped as a {@link Deque}
+    * @return a new {@link MarshallableDeque} instance containing the passed object if the object is not null,
+    * otherwise null. If the passed queue is not a {@link Deque} implementation, a new one is created.
+    */
+   public static <T> MarshallableDeque<T> create(Queue<T> queue) {
+      return queue == null ? null :
+            queue instanceof Deque<T> deque ?
```

Comment:
are we using Deque?

Reply:
We actually only need a `Queue` implementation for `CacheStreamIntermediateReducer`, but I implemented it as `Deque` to give us more flexibility in the future if needed.

Reply:
@wburns why is it a `Queue`? Can't it be a `List` as it seems only to invoke `add`?

Reply:
It can be, it was just Deque since it didn't need to have random access when I wrote this many years ago :) The code was changed to use Iterable anyways though afaik.

While I agree we can probably wait on supporting Queue, I don't think it hurts either.

Reply:
Unfortunately even with my Protostream enhancements we can't use `Iterable` directly in this case because we need to marshall an interface `IntermediateOperation`, not a concrete type.

Assuming `IntermediateOperation` implementations are always provided by us, we could do something like:

```java
   @ProtoField(number = 1)
   Stream<WrappedMessage> getIntermediateOperations() {
      return intOps.stream().map(WrappedMessage::new);
   }
```
The disadvantage being that we're creating a `WrappedMesssage` per op, which is avoided when using the various `Marshallable*` abstractions.